File: src/trust-memory.ts
/**
 * Seal Gate v0.4 — Trust Memory
 * Cross-session agent reliability tracking: recurrent pattern count,
 * reliability score, and drift trend.
 *
 * Stateless core: TrustMemory holds in-memory state.
 * Callers are responsible for persistence (serialize/deserialize via toJSON/fromJSON).
 */
import { Verdict } from './types.ts'

export interface ReviewRecord {
  ts: number           // Unix ms
  verdict: Verdict
  trust_score: number  // 0–100 final trust_score from Seal.review()
  risk_level: string
  blocking_count: number
}

export interface AgentSummary {
  agent_id: string
  reliability_score: number | null  // null if no history
  drift_trend: 'improving' | 'stable' | 'degrading'
  review_count: number
  last_reviewed_at: number | null
}

export interface TrustMemoryOptions {
  maxHistory?: number  // cap per agent (default 50)
  driftThreshold?: number  // score delta to classify as improving/degrading (default 15)
}

// Verdict → base score contribution (0–100)
const VERDICT_SCORE: Record<Verdict, number> = {
  'PASS': 100,
  'PASS_WITH_WARNINGS': 80,
  'REVISE': 50,
  'ESCALATE_TO_HUMAN': 20,
  'BLOCK': 0,
}

/**
 * Compute weighted reliability score from review history.
 * More recent reviews have higher weight (linear decay).
 * Returns null for empty history.
 */
export function computeReliabilityScore(records: ReviewRecord[]): number | null {
  if (records.length === 0) return null

  const n = records.length
  let totalWeight = 0
  let weightedSum = 0

  records.forEach((r, i) => {
    // Linear weight: oldest = 1, most recent = n
    const weight = i + 1
    const verdictScore = VERDICT_SCORE[r.verdict] ?? 50
    // Blend verdict score (70%) with actual trust_score (30%)
    const blended = verdictScore * 0.7 + r.trust_score * 0.3
    weightedSum += blended * weight
    totalWeight += weight
  })

  return Math.round(weightedSum / totalWeight)
}

/**
 * Classify drift trend by splitting at the time midpoint (not count midpoint).
 * Time-midpoint avoids skew when reviews are temporally clustered.
 * Returns 'stable' for fewer than 2 records or when either half is empty.
 */
export function getDriftTrend(
  records: ReviewRecord[],
  threshold = 15,
): 'improving' | 'stable' | 'degrading' {
  if (records.length < 2) return 'stable'

  // Sort by ts to ensure correct temporal split
  const sorted = [...records].sort((a, b) => a.ts - b.ts)
  const midTs = (sorted[0].ts + sorted[sorted.length - 1].ts) / 2
  const first  = sorted.filter(r => r.ts <= midTs)
  const second = sorted.filter(r => r.ts > midTs)

  if (first.length === 0 || second.length === 0) return 'stable'

  const avg = (rs: ReviewRecord[]) =>
    rs.reduce((s, r) => s + r.trust_score, 0) / rs.length

  const delta = avg(second) - avg(first)

  if (delta > threshold) return 'improving'
  if (delta < -threshold) return 'degrading'
  return 'stable'
}

/** Sort by ts and cap — used for real-time record(). */
function sortAndCap(records: ReviewRecord[], maxHistory: number): ReviewRecord[] {
  return [...records].sort((a, b) => a.ts - b.ts).slice(-maxHistory)
}

/** Sort, dedup same-ts (keep last), and cap — used only when loading persisted data. */
function normalizeFromPersistence(records: ReviewRecord[], maxHistory: number): ReviewRecord[] {
  const sorted = [...records].sort((a, b) => a.ts - b.ts)
  const seen = new Map<number, ReviewRecord>()
  for (const r of sorted) seen.set(r.ts, r)
  return [...seen.values()].sort((a, b) => a.ts - b.ts).slice(-maxHistory)
}

export class TrustMemory {
  private store: Map<string, ReviewRecord[]> = new Map()
  private readonly maxHistory: number
  private readonly driftThreshold: number

  constructor(options: TrustMemoryOptions = {}) {
    this.maxHistory = options.maxHistory ?? 50
    this.driftThreshold = options.driftThreshold ?? 15
  }

  /** Append a review result for an agent. Keeps history sorted by ts, capped to maxHistory. */
  record(agent_id: string, review: Omit<ReviewRecord, 'ts'> & { ts?: number }): void {
    const entry: ReviewRecord = { ts: review.ts ?? Date.now(), ...review }
    const history = this.store.get(agent_id) ?? []
    history.push(entry)
    this.store.set(agent_id, sortAndCap(history, this.maxHistory))
  }

  /** Get full review history for an agent (oldest first). */
  getHistory(agent_id: string): ReviewRecord[] {
    return [...(this.store.get(agent_id) ?? [])]
  }

  /** Reliability score 0–100, or null if no history. */
  getScore(agent_id: string): number | null {
    return computeReliabilityScore(this.store.get(agent_id) ?? [])
  }

  /** Full summary for an agent. */
  getSummary(agent_id: string): AgentSummary {
    const history = this.store.get(agent_id) ?? []
    const last = history.at(-1)
    return {
      agent_id,
      reliability_score: computeReliabilityScore(history),
      drift_trend: getDriftTrend(history, this.driftThreshold),
      review_count: history.length,
      last_reviewed_at: last?.ts ?? null,
    }
  }

  /** Serialize to plain object (for persistence). */
  toJSON(): Record<string, ReviewRecord[]> {
    return Object.fromEntries(this.store)
  }

  /** Restore from serialized object. Validates, sorts by ts, deduplicates, and caps. */
  static fromJSON(data: Record<string, ReviewRecord[]>, options?: TrustMemoryOptions): TrustMemory {
    const mem = new TrustMemory(options)
    const VALID_VERDICTS = new Set(['PASS', 'PASS_WITH_WARNINGS', 'REVISE', 'ESCALATE_TO_HUMAN', 'BLOCK'])
    for (const [agent_id, records] of Object.entries(data)) {
      if (!Array.isArray(records)) continue
      const valid = records.filter(r =>
        typeof r.ts === 'number' &&
        typeof r.verdict === 'string' && VALID_VERDICTS.has(r.verdict) &&
        typeof r.trust_score === 'number' &&
        typeof r.risk_level === 'string' &&
        typeof r.blocking_count === 'number'
      )
      mem.store.set(agent_id, normalizeFromPersistence(valid, options?.maxHistory ?? 50))
    }
    return mem
  }
}

File: python/seal_gate/trust_memory.py
"""
Seal Gate v0.4 — Trust Memory
Cross-session agent reliability tracking: reliability score and drift trend.
Callers handle persistence via to_dict/from_dict.
"""
from __future__ import annotations
import math
import time
from dataclasses import dataclass, field
from typing import Optional

VERDICT_SCORE: dict[str, int] = {
    'PASS': 100,
    'PASS_WITH_WARNINGS': 80,
    'REVISE': 50,
    'ESCALATE_TO_HUMAN': 20,
    'BLOCK': 0,
}


@dataclass
class ReviewRecord:
    ts: int  # Unix ms
    verdict: str
    trust_score: int
    risk_level: str
    blocking_count: int

    def to_dict(self) -> dict:
        return {'ts': self.ts, 'verdict': self.verdict, 'trust_score': self.trust_score,
                'risk_level': self.risk_level, 'blocking_count': self.blocking_count}

    @staticmethod
    def from_dict(d: dict) -> 'ReviewRecord':
        return ReviewRecord(ts=d['ts'], verdict=d['verdict'], trust_score=d['trust_score'],
                            risk_level=d['risk_level'], blocking_count=d['blocking_count'])


@dataclass
class AgentSummary:
    agent_id: str
    reliability_score: Optional[int]
    drift_trend: str  # 'improving' | 'stable' | 'degrading'
    review_count: int
    last_reviewed_at: Optional[int]


def compute_reliability_score(records: list[ReviewRecord]) -> Optional[int]:
    """Weighted reliability score (recent reviews weighted higher). Returns None if no records."""
    if not records:
        return None
    n = len(records)
    total_weight = 0
    weighted_sum = 0.0
    for i, r in enumerate(records):
        weight = i + 1  # linear: oldest=1, most recent=n
        verdict_score = VERDICT_SCORE.get(r.verdict, 50)
        blended = verdict_score * 0.7 + r.trust_score * 0.3
        weighted_sum += blended * weight
        total_weight += weight
    # Use half-up rounding to match JS Math.round() (Python uses banker's rounding by default)
    return math.floor(weighted_sum / total_weight + 0.5)


def get_drift_trend(records: list[ReviewRecord], threshold: int = 15) -> str:
    """Split at time midpoint (not count midpoint) to avoid skew with clustered reviews."""
    if len(records) < 2:
        return 'stable'
    sorted_r = sorted(records, key=lambda r: r.ts)
    mid_ts = (sorted_r[0].ts + sorted_r[-1].ts) / 2
    first  = [r for r in sorted_r if r.ts <= mid_ts]
    second = [r for r in sorted_r if r.ts > mid_ts]
    if not first or not second:
        return 'stable'

    def avg(rs: list[ReviewRecord]) -> float:
        return sum(r.trust_score for r in rs) / len(rs)

    delta = avg(second) - avg(first)
    if delta > threshold:
        return 'improving'
    if delta < -threshold:
        return 'degrading'
    return 'stable'


def _sort_and_cap(records: list[ReviewRecord], max_history: int) -> list[ReviewRecord]:
    """Sort by ts and cap — for real-time record()."""
    return sorted(records, key=lambda r: r.ts)[-max_history:]


def _normalize_from_persistence(records: list[ReviewRecord], max_history: int) -> list[ReviewRecord]:
    """Sort, dedup same-ts (keep last), cap — for loading persisted data only."""
    sorted_r = sorted(records, key=lambda r: r.ts)
    seen: dict[int, ReviewRecord] = {}
    for r in sorted_r:
        seen[r.ts] = r
    return sorted(seen.values(), key=lambda r: r.ts)[-max_history:]


VALID_VERDICTS = {'PASS', 'PASS_WITH_WARNINGS', 'REVISE', 'ESCALATE_TO_HUMAN', 'BLOCK'}


class TrustMemory:
    def __init__(self, max_history: int = 50, drift_threshold: int = 15):
        self._store: dict[str, list[ReviewRecord]] = {}
        self._max_history = max_history
        self._drift_threshold = drift_threshold

    def record(self, agent_id: str, verdict: str, trust_score: int,
               risk_level: str, blocking_count: int, ts: Optional[int] = None) -> None:
        if ts is not None:
            effective_ts = ts
        else:
            # Ensure monotonically increasing ts even within the same millisecond
            now_ms = int(time.time() * 1000)
            history = self._store.get(agent_id, [])
            last_ts = history[-1].ts if history else 0
            effective_ts = max(now_ms, last_ts + 1)
        r = ReviewRecord(ts=effective_ts, verdict=verdict,
                         trust_score=trust_score, risk_level=risk_level, blocking_count=blocking_count)
        history = self._store.setdefault(agent_id, [])
        history.append(r)
        self._store[agent_id] = _sort_and_cap(history, self._max_history)

    def get_history(self, agent_id: str) -> list[ReviewRecord]:
        return list(self._store.get(agent_id, []))

    def get_score(self, agent_id: str) -> Optional[int]:
        return compute_reliability_score(self._store.get(agent_id, []))

    def get_summary(self, agent_id: str) -> AgentSummary:
        history = self._store.get(agent_id, [])
        return AgentSummary(
            agent_id=agent_id,
            reliability_score=compute_reliability_score(history),
            drift_trend=get_drift_trend(history, self._drift_threshold),
            review_count=len(history),
            last_reviewed_at=history[-1].ts if history else None,
        )

    def to_dict(self) -> dict:
        return {agent_id: [r.to_dict() for r in records] for agent_id, records in self._store.items()}

    @staticmethod
    def from_dict(data: dict, max_history: int = 50, drift_threshold: int = 15) -> 'TrustMemory':
        mem = TrustMemory(max_history=max_history, drift_threshold=drift_threshold)
        for agent_id, records in data.items():
            if not isinstance(records, list):
                continue
            valid = []
            for r in records:
                if not isinstance(r, dict):
                    continue
                if not (isinstance(r.get('ts'), (int, float)) and
                        r.get('verdict') in VALID_VERDICTS and
                        isinstance(r.get('trust_score'), (int, float)) and
                        isinstance(r.get('risk_level'), str) and
                        isinstance(r.get('blocking_count'), (int, float))):
                    continue
                valid.append(ReviewRecord.from_dict(r))
            if valid:
                mem._store[agent_id] = _normalize_from_persistence(valid, max_history)
        return mem
